File: VB/WindowsApplication3/CheckedImageComboBoxEditPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.Skins
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports System.Drawing
Imports System.Reflection
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Utils.Text


Namespace DXSample

	Public Class CheckedImageComboBoxEditPainter
		Inherits ButtonEditPainter
		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Sub DrawString(ByVal info As ControlGraphicsInfoArgs, ByVal bounds As Rectangle, ByVal text As String, ByVal appearance As DevExpress.Utils.AppearanceObject)
			Dim viewInfo As PopupContainerEditViewInfo = TryCast(info.ViewInfo, PopupContainerEditViewInfo)
			Dim edit As RepositoryItemCheckedImageComboBoxEdit = TryCast(viewInfo.Item, RepositoryItemCheckedImageComboBoxEdit)
			If (Not edit.CanShowImageInEditBox) Then
				MyBase.DrawString(info, bounds, text, appearance)
			Else
                DrawEditContent(info, bounds, text, appearance)
            End If
        End Sub

        Private Sub DrawEditContent(ByVal info As ControlGraphicsInfoArgs, ByVal bounds As Rectangle, ByVal text As String, ByVal appearance As DevExpress.Utils.AppearanceObject)
            Dim viewInfo As PopupContainerEditViewInfo = TryCast(info.ViewInfo, PopupContainerEditViewInfo)
            Dim edit As RepositoryItemCheckedImageComboBoxEdit = TryCast(viewInfo.Item, RepositoryItemCheckedImageComboBoxEdit)
            Dim values() As String = text.Split(edit.SeparatorChar)
            Dim valRect As New Rectangle(bounds.X, bounds.Y, 0, bounds.Height)
            For Each val As String In values
                Dim index As Integer = GetIndexByDescription(edit.GetItems(), val.Trim())
                If index <> -1 Then
                    DrawImage(edit, info.Graphics, index, valRect)

                End If
                Dim str As String = val
                If Array.IndexOf(values, val) < values.Length - 1 Then
                    str = String.Concat(val, edit.SeparatorChar)
                End If
                DrawDisplayText(info, appearance, valRect, str)
            Next val
        End Sub

        Private Sub DrawDisplayText(ByVal info As ControlGraphicsInfoArgs, ByVal appearance As DevExpress.Utils.AppearanceObject, ByRef valRect As Rectangle, ByVal str As String)
            Dim width As Integer = TextUtils.GetStringSize(info.Graphics, str, appearance.Font).Width
            valRect.Width = width
            MyBase.DrawString(info, valRect, str, appearance)
            valRect.X += width
        End Sub

		Private Sub DrawImage(ByVal edit As RepositoryItemCheckedImageComboBoxEdit, ByVal gr As Graphics, ByVal index As Integer, ByRef valRect As Rectangle)
			Dim image As Image = edit.GetItemImage(index)
			Dim rect As New Rectangle(valRect.X, valRect.Y, RepositoryItemCheckedImageComboBoxEdit.ImageWidth, RepositoryItemCheckedImageComboBoxEdit.ImageHeight)
			If image IsNot Nothing Then
				gr.DrawImage(image, rect)
				valRect.X += RepositoryItemCheckedImageComboBoxEdit.ImageWidth + 2
			End If
		End Sub

		Private Function GetIndexByDescription(ByVal collection As CheckedListBoxItemCollection, ByVal description As String) As Integer
			For i As Integer = 0 To collection.Count - 1
				If collection(i).Description.Equals(description) Then
					Return i
				End If
			Next i
			Return -1
		End Function
	End Class
End Namespace

File: CS/WindowsApplication3/CheckedImageComboBoxEditPainter.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.Skins;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using System.Drawing;
using System.Reflection;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Controls;
using DevExpress.Utils.Text;


namespace DXSample {

    public class CheckedImageComboBoxEditPainter: ButtonEditPainter {
        public CheckedImageComboBoxEditPainter() : base() { }

        protected override void DrawString(ControlGraphicsInfoArgs info, Rectangle bounds, string text, DevExpress.Utils.AppearanceObject appearance) {
            PopupContainerEditViewInfo viewInfo = info.ViewInfo as PopupContainerEditViewInfo;
            RepositoryItemCheckedImageComboBoxEdit edit = viewInfo.Item as RepositoryItemCheckedImageComboBoxEdit;
            if(!edit.CanShowImageInEditBox)
                base.DrawString(info, bounds, text, appearance);
            else 
                DrawContent(info, bounds, text, appearance);
        }

        private void DrawContent(ControlGraphicsInfoArgs info, Rectangle bounds, string text, DevExpress.Utils.AppearanceObject appearance) {
            PopupContainerEditViewInfo viewInfo = info.ViewInfo as PopupContainerEditViewInfo;
            RepositoryItemCheckedImageComboBoxEdit edit = viewInfo.Item as RepositoryItemCheckedImageComboBoxEdit;
            string[] values = text.Split(edit.SeparatorChar);
            Rectangle valRect = new Rectangle(bounds.X, bounds.Y, 0, bounds.Height);
            foreach(string val in values) {
                int index = GetIndexByDescription(edit.GetItems(), val.Trim());
                if(index != -1) {
                    DrawImage(edit, info.Graphics, index, ref valRect);
                   
                }
                string str = val;
                if(Array.IndexOf(values, val) < values.Length - 1)
                    str = string.Concat(val, edit.SeparatorChar);
                DrawString(info, appearance, ref valRect, str);
            }
        }

        private void DrawString(ControlGraphicsInfoArgs info, DevExpress.Utils.AppearanceObject appearance, 
            ref Rectangle valRect, string str) {
            int width = TextUtils.GetStringSize(info.Graphics, str, appearance.Font).Width;
            valRect.Width = width;
            base.DrawString(info, valRect, str, appearance);
            valRect.X += width;
        }

        private void DrawImage(RepositoryItemCheckedImageComboBoxEdit edit, Graphics gr, int index, ref Rectangle valRect) {
            Image image = edit.GetItemImage(index);
            Rectangle rect = new Rectangle(valRect.X, valRect.Y, RepositoryItemCheckedImageComboBoxEdit.ImageWidth,
                RepositoryItemCheckedImageComboBoxEdit.ImageHeight);
            if(image != null) {
                gr.DrawImage(image, rect);
                valRect.X += RepositoryItemCheckedImageComboBoxEdit.ImageWidth + 2;
            }
        }

        int GetIndexByDescription(CheckedListBoxItemCollection collection, string description) {
            for(int i = 0;i < collection.Count;i++) {
                if(collection[i].Description.Equals(description))
                    return i;
            }
            return -1;
        }
    }
}


